docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md

```diff
@@ -78,15 +78,149 @@ If you need to have external labels in your query results, this is currently not
-> Please refer to [`thanos tools bucket rewrite` documentation](https://thanos.io/tip/components/tools.md/#bucket-rewrite) for more details.
-
 **Grafana Mimir does not support Thanos’ _downsampling_ feature.**
 To guarantee query results correctness please only upload original (raw) Thanos blocks into Mimir's storage.
 If you also upload blocks with downsampled data (ie. blocks with non-zero `Resolution` field in `meta.json` file), Grafana Mimir will merge raw samples and downsampled samples together at the query time.
 This may cause that incorrect results are returned for the query.
 
-> **Note**: It is possible to run compaction and deduplicate blocks by Thanos first by using `thanos compact` command
-> with `--compact.enable-vertical-compaction --deduplication.func=penalty --deduplication.replica-label=<LABEL>` flags.
-> Please refer to [Vertical Compaction Use Cases](https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases) for more details on offline deduplication.
+## Migrating historic TSDB blocks from Thanos to Grafana Mimir
+
+1) Copy the blocks from Thanos's bucket to an Intermediate bucket
+
+   Create an intermediate object storage bucket(S3,GCS,etc.) in your cloud provider where you can copy the historical blocks and work on them before uploading to the Mimir bucket.
+
+    For AWS S3, use the aws tool in the following command:
+
+    ```bash
+    aws s3 cp -r s3://<THANOS-BUCKET> s3://<TENANT>/<DIRECTORY>
+    ```
+
+    For Google Cloud Storage (GCS), use the gsutil tool in the following command:
+    
+    ```bash
+    gsutil -m cp -r gs://<THANOS-BUCKET>/* gs://<INTERMEDIATE-MIMIR-BUCKET>/
+    ```
+    
+    Once the copy process is done, inspect the blocks in the bucket to ensure it's valid from Thanos perspective.
+    
+    ```bash
+    thanos tools bucket inspect --objstore.config-file bucket.yaml
+    ```
+
+2) Remove the downsampled blocks
+
+   Mimir doesn't understand the downsampled blocks(ie. blocks with non-zero Resolution field in meta.json file) in Thanos. Therefore we need to remove the 5m and 1h downsampled blocks from this bucket.
+   
+   Mark the downsampled blocks for deletion.
+   
+   ```bash
+   thanos tools bucket retention --objstore.config-file bucket.yaml --retention.resolution-1h=2h --retention.resolution-5m=2h --retention.resolution-raw=0d
+   ```
+   Cleanup the blocks marked for deletion.
+   
+   ```bash
+   thanos tools bucket cleanup --objstore.config-file bucket.yaml --delete-delay=0
+   ```
+  
+3) Remove the duplicated blocks
+
+   If two replicas of Prometheus instancesare deployed for HA, then only upload the blocks from one of the replicas and drop the other.
+   
+   ```bash
+   thanos tools bucket inspect --objstore.config-file bucket.yaml --output=tsv | grep prometheus_replica=<PROMETHEUS-REPLICA-TO-DROP> | awk '{print $1}' | xargs -n 1 -P 1 thanos tools bucket mark --marker="deletion-mark.json" --objstore.config-file bucket.yaml --details="Removed as duplicate" --id
```

Comment:
This command combines several steps into one long pipeline.

Can we split it into multiple commands with a comment? Something like:

```bash
# Get list of all blocks in the bucket
thanos tools bucket inspect --objstore.config-file bucket.yaml --output=tsv > blocks.tsv

# Find blocks from replica that we will drop
cat blocks.tsv | grep prometheus_replica=<PROMETHEUS-REPLICA-TO-DROP> | awk '{print $1}' > blocks_to_drop.tsv

# Mark found blocks for deletion
for ID in $(cat blocks_to_drop.tsv); do
  thanos tools bucket mark --marker="deletion-mark.json" --objstore.config-file bucket.yaml --details="Removed as duplicate" --id=$ID
done
```

Reply:
We can split the command but the thing with "for" loop is that they tend to process one item at a time and this is where xargs has an advantage - it can batch up the arguments to allow one command to process lots of items. Also when we go for loop here we will have to maintain so many files as well. May be I can give a small description of what the command does. Will that be okay rather than splitting the command?

Reply:
`xargs -n 1` will call the command once per block anyway, right? What is the difference then?

Reply:
Yeah in this command, we are not using any batching since the max procs we are using is 1. If you feel that would be easier to follow then let's adopt that.